File: packages/sdk-typescript/src/span.ts
import { SpanStatusCode } from '@opentelemetry/api'
import type { Span as OtelSpan } from '@opentelemetry/api'
import type { SpanEndOptions } from './types'

export interface TracionSpan {
  setInput(value: unknown): void
  setOutput(value: unknown): void
  setAttribute(key: string, value: unknown): void
  addEvent(name: string, attributes?: Record<string, unknown>): void
  end(options?: SpanEndOptions): void
}

export class OtelTracionSpan implements TracionSpan {
  constructor(private readonly otelSpan: OtelSpan) {}

  setInput(value: unknown): void {
    this.otelSpan.setAttribute('tracion.input', JSON.stringify(value))
  }

  setOutput(value: unknown): void {
    this.otelSpan.setAttribute('tracion.output', JSON.stringify(value))
  }

  setAttribute(key: string, value: unknown): void {
    this.otelSpan.setAttribute(key, JSON.stringify(value))
  }

  addEvent(name: string, attributes?: Record<string, unknown>): void {
    this.otelSpan.addEvent(name, attributes as Record<string, string | number | boolean>)
  }

  end(options?: SpanEndOptions): void {
    if (options?.status === 'error') {
      const message = options.error instanceof Error
        ? options.error.message
        : String(options.error ?? '')
      this.otelSpan.setStatus({ code: SpanStatusCode.ERROR, message })
    } else {
      this.otelSpan.setStatus({ code: SpanStatusCode.OK })
    }
    this.otelSpan.end()
  }
}

export class NoopTracionSpan implements TracionSpan {
  setInput(_value: unknown): void { /* no-op */ }
  setOutput(_value: unknown): void { /* no-op */ }
  setAttribute(_key: string, _value: unknown): void { /* no-op */ }
  addEvent(_name: string, _attributes?: Record<string, unknown>): void { /* no-op */ }
  end(_options?: SpanEndOptions): void { /* no-op */ }
}

File: packages/sdk-python/src/tracion/_span.py
import json
from abc import ABC, abstractmethod
from typing import Any, cast

from opentelemetry.trace import Span as OtelSpan, StatusCode, Status
from opentelemetry.util.types import Attributes

from tracion._types import TraceStatus


class TracionSpan(ABC):
    @abstractmethod
    def set_input(self, value: Any) -> None: ...

    @abstractmethod
    def set_output(self, value: Any) -> None: ...

    @abstractmethod
    def set_attribute(self, key: str, value: Any) -> None: ...

    @abstractmethod
    def add_event(self, name: str, attributes: dict[str, Any] | None = None) -> None: ...

    @abstractmethod
    def end(self, status: TraceStatus = "success", error: BaseException | None = None) -> None: ...


class OtelTracionSpan(TracionSpan):
    def __init__(self, otel_span: OtelSpan) -> None:
        self._otel_span = otel_span

    def set_input(self, value: Any) -> None:
        self._otel_span.set_attribute("tracion.input", json.dumps(value))

    def set_output(self, value: Any) -> None:
        self._otel_span.set_attribute("tracion.output", json.dumps(value))

    def set_attribute(self, key: str, value: Any) -> None:
        if isinstance(value, (str, bool, int, float)):
            self._otel_span.set_attribute(key, value)
        else:
            self._otel_span.set_attribute(key, json.dumps(value))

    def add_event(self, name: str, attributes: dict[str, Any] | None = None) -> None:
        self._otel_span.add_event(name, cast(Attributes, attributes) if attributes else None)

    def end(self, status: TraceStatus = "success", error: BaseException | None = None) -> None:
        if status == "error":
            message = str(error) if error else ""
            self._otel_span.set_status(Status(StatusCode.ERROR, message))
        else:
            self._otel_span.set_status(Status(StatusCode.OK))
        self._otel_span.end()


class NoopTracionSpan(TracionSpan):
    def set_input(self, value: Any) -> None:
        pass

    def set_output(self, value: Any) -> None:
        pass

    def set_attribute(self, key: str, value: Any) -> None:
        pass

    def add_event(self, name: str, attributes: dict[str, Any] | None = None) -> None:
        pass

    def end(self, status: TraceStatus = "success", error: BaseException | None = None) -> None:
        pass
